src/nlr/MILPFormulator.cpp

```diff
@@ -0,0 +1,486 @@
+            storeUbIfNeeded( layer, i, variable, newUb );
+            if ( _cutoffInUse && newUb < _cutoffValue )
+            {
+                ++_cutoffs;
+                continue;
+            }
+        }
+    }
+
+    struct timespec gurobiEnd = TimeUtils::sampleMicro();
+
+    log( Stringf( "Number of tighter bounds found by Gurobi: %u. Sign changes: %u. Cutoffs: %u\n",
+                  _tighterBoundCounter, _signChanges, _cutoffs ) );
+    log( Stringf( "Seconds spent Gurobiing: %llu\n", TimeUtils::timePassed( gurobiStart, gurobiEnd ) / 1000000 ) );
+}
+
+void MILPFormulator::createMILPEncoding( const Map<unsigned, Layer *> &layers,
+                                         GurobiWrapper &gurobi,
+                                         unsigned lastLayer )
+{
+    // First, create the LP relaxation of the problem
+    _lpFormulator.createLPRelaxation( layers, gurobi, lastLayer );
+
+    // Now, add the MILP constraints
+    for ( const auto &layer : layers )
+    {
+        if ( layer.second->getLayerIndex() > lastLayer )
+            continue;
+
+        addLayerToModel( gurobi, layer.second );
+    }
+}
+
+void MILPFormulator::addLayerToModel( GurobiWrapper &gurobi, const Layer *layer )
+{
+    switch ( layer->getLayerType() )
+    {
+        case Layer::INPUT:
+        case Layer::WEIGHTED_SUM:
+            break;
+
+        case Layer::RELU:
+            addReluLayerToMILPFormulation( gurobi, layer );
+            break;
+
+        default:
+            throw NLRError( NLRError::LAYER_TYPE_NOT_SUPPORTED, "MILPFormulator" );
+            break;
+    }
+}
+
+void MILPFormulator::addReluLayerToMILPFormulation( GurobiWrapper &gurobi,
+                                                    const Layer *layer )
+{
+    for ( unsigned i = 0; i < layer->getSize(); ++i )
+    {
+        if ( !layer->neuronEliminated( i ) )
+        {
+            unsigned targetVariable = layer->neuronToVariable( i );
+
+            List<NeuronIndex> sources = layer->getActivationSources( i );
+            const Layer *sourceLayer = _layerOwner->getLayer( sources.begin()->_layer );
+            unsigned sourceNeuron = sources.begin()->_neuron;
+            unsigned sourceVariable = sourceLayer->neuronToVariable( sourceNeuron );
+
+            double sourceLb = sourceLayer->getLb( sourceNeuron );
+            double sourceUb = sourceLayer->getUb( sourceNeuron );
+
+            if ( !FloatUtils::isNegative( sourceLb ) ||
+                 !FloatUtils::isPositive( sourceUb ) )
+            {
+                // This ReLU is fixed on one of its phases
+                continue;
+            }
+
+            /*
+              The underlying LP relaxation defines the triangular
+              region; we add the indicator variable a \in {0,1}:
+
+                y <= x - l ( 1 - a )
+                y <= u a
+
+                Or, alternatively:
+
+                y - x -la <= -l
+                y - ua <= 0
+            */
```

Comment:
Would the constraints
y >= 0 and y <= x be captured?

Reply:
Yes, these constraints are part of the underlying LP triangular constraints, which are added earlier by the lpFormulator